nutils/types.py

```diff
@@ -1481,6 +1481,46 @@ def __setitem__(self, item, value):
     super().__setitem__(item, value)
     self.__keys.appendleft(item)
 
+def lru_cache(func=None, maxsize=None):
+  '''Buffer-aware LRU cache.
+
+  Return values are cached for immutable arguments are cached, otherwise
+  caching is silently disabled. Any array arguments are equality-tested based
+  on the underlying buffer, with an callback placed on the memory release that
+  removes the corresponding cache entry.
+  '''
+
+  if func is None:
+    return functools.partial(lru_cache, maxsize=maxsize)
+
+  cache = {} if maxsize is None else lru_dict(maxsize)
+
+  @functools.wraps(func)
+  def wrapper(*args):
+    key = []
+    for arg in args:
+      if not _isfrozen(arg):
+        return func(*args)
+      key.append(Py_buffer(arg) if isinstance(arg, numpy.ndarray) else (type(arg), arg))
+    key = tuple(key)
+    try:
+      v = cache[key]
+    except KeyError:
+      v = cache[key] = func(*args)
+      _deepfreeze(v)
+      pop = functools.partial(cache.pop, key)
+      for buf in key:
+        if isinstance(buf, Py_buffer):
+          # replace the buffer exporter by a weak reference of the lowest base,
+          # with a callback at destruction to remove the cache entry
+          obj = buf.obj
+          while isinstance(obj, numpy.ndarray) and obj.base is not None:
+            obj = obj.base
+          buf.replace_obj(weakref.ref(obj, pop))
+    return v
+
```

Comment:
We could make `cache` an attribute of `wrapper` to enable better debugging.

Reply:
I experimented with making `lru_cache` a class but it added noticeable overhead, probably due to the logic required for correctly handling bound methods. Instead I'm thinking in the direction of adding caching dictionaries to a global registery (including those of `typing.CacheMeta`) so we have a single place to search for memory leaks.

Reply:
What I meant is actually just put `wrapper._cache = cache` before return so cache is exposed and always link to the cached function.
Regarding a single point for debugging, I guess it's easy enough to also attach a UUID to the function(is a function object hashable?) and use that as key and attach them to a global registry: `{uuid: (f, f._cache)}` or simply `{uuid: f}`.

Reply:
Check, that's indeed a lot easier. Added as `.cache`.